compiler-rt/lib/rtsan/rtsan_interceptors_posix.cpp

```diff
@@ -464,10 +464,13 @@ INTERCEPTOR(void *, valloc, SIZE_T size) {
 }
 
 #if SANITIZER_INTERCEPT_ALIGNED_ALLOC
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wunguarded-availability-new"
```

Comment:
Put the "Why is this OK? (to disable this warning)" here as a comment.

Reply:
Good suggestion, added a more succinct version but kept the spirit of it alive.

Reply:
hmm, why does it link then though?

Reply:
(It doesn't seem `REAL(...)` is a weak function for Apple, but I don't really know details about how things work there)

Reply:
The SDK being used while linking _does_ have aligned_alloc (in System.B.dylib), so it will link fine - but attempt to run the resulting binary on macOS 10.13 will result in a crash. Hence the warning (converted to error in Apple fork, so it won't build with -mminmacos=10.13)

Reply:
sorry, for my education: i meant the dynamic loader, does that just ignore unknown symbols (as it would for ELF if they are weak?)

Reply:
From what I see in the comments around SANITIZER_WEAK_IMPORT in  different places (TSAN sources) (aligned_alloc is decorated with it), it might be that the call to the function will be 'ignored'. Perhaps there'll be some default (empty) impl of it, so it won't perhaps crash. But also it won't do what the function normally is expected to do, so the crash might still happen ('default' aligned_alloc might return NULL?)

Sorry, I'm not very familiar with these weak references, so perhaps my understanding is a bit vague.

Reply:
https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPFrameworks/Concepts/WeakLinking.html

This documentation may be helpful.


My understanding below:

If code that is **built** on a build machine that is aware of `aligned_alloc`, and the binary has a function that calls it "unguarded":

```cpp
int main() 
{
    aligned_alloc(alignof(int), sizeof(int)); 
}

```

It will crash when it **runs** on systems that do not have this function defined (OSX < 10.15)

You can get around this by guarding it:
```
int main() 
{
    if (__builtin_available(macOS 10.15, *))
        aligned_alloc(alignof(int), sizeof(int)); 
    else 
       malloc(sizeof(int));
}
```

In this PR we add aligned_alloc as a weak import so it will link on our build machine, if the build machine is 10.15 or greater (aware of the symbol `__MAC_10_15`). This is necessary because the OS version we build for is min version 10.13.

We define the interceptor function, and turn of the unguarded warning, so if a client machine has `aligned_alloc` it will be intercepted.

If the client machine does not have `aligned_alloc` (<10.15), it will not have `aligned_alloc` (a little circular logic, for you), and we will never intercept anything, so us defining this interceptor will have no effect and cause no issue

Reply:
maybe my messages were too confusing, I meant: is this weakly-linked? If so please mention in the comment.

Reply:
Let me know if that new comment works for you - I'm a newbie to weakly/strongly linked symbols, so I am open to alternate wordings if this one doesn't make sense.

Reply:
Hmm. I don't really like the wording, the weak consideration is kind of separate. something like

"Because we are providing a wrapper, the client is responsible to check whether aligned_alloc is available, not us.
We still succeed linking on an old OS, because we are using a weak symbol."

Reply:
Take 3 is up. Thanks for the help on this one.